intake_xarray/image.py

```diff
@@ -0,0 +1,265 @@
+        part of the the pattern.
+    xarray_kwargs : dict, optional
+        Any further arguments to pass to reader function.
+    """
+    name = 'xarray_image'
+
+    def __init__(self, urlpath, chunks=None, concat_dim='concat_dim',
+                 xarray_kwargs=None, metadata=None, path_as_pattern=True,
+                 storage_options=None, **kwargs):
+        self.path_as_pattern = path_as_pattern
+        self.urlpath = urlpath
+        self.chunks = chunks
+        self.concat_dim = concat_dim
+        self.storage_options = storage_options or {}
+        self._kwargs = xarray_kwargs or kwargs
+        self._ds = None
+        super(ImageSource, self).__init__(metadata=metadata)
+
+    def _open_files(self, files):
+        import pandas as pd
+        from xarray import DataArray
+
+        out = multireader(files, self.chunks, self.concat_dim, **self._kwargs)
+        if not self.pattern:
+            return out
+
+        coords = {}
+        filenames = [f.path for f in files]
+        field_values = reverse_formats(self.pattern, filenames)
+
+        if isinstance(self.concat_dim, list):
+            if not set(field_values.keys()).issuperset(set(self.concat_dim)):
+                raise KeyError('All concat_dims should be in pattern.')
+            index = pd.MultiIndex.from_tuples(
+                zip(*(field_values[dim] for dim in self.concat_dim)),
+                names=self.concat_dim)
+            coords = {
+                k: DataArray(v, dims=('dim_0'))
+                for k, v in field_values.items() if k not in self.concat_dim
+            }
+            out = (out.assign_coords(dim_0=index, **coords)  # use the index
+                      .unstack().chunk(self.chunks))  # unstack along new index
+            return out.transpose(*self.concat_dim,  # reorder dims
+                                 *filter(lambda x: x not in self.concat_dim,
+                                         out.dims))
+        else:
+            coords = {
+                k: DataArray(v, dims=self.concat_dim)
+                for k, v in field_values.items()
+            }
+            return out.assign_coords(**coords).chunk(self.chunks)
+
+    def _open_dataset(self):
+        from dask.bytes import open_files
+
+        files = open_files(self.urlpath, **self.storage_options)
+        if len(files) == 0:
+            raise Exception("No files found at {}".format(self.urlpath))
+        if len(files) == 1:
+            self._ds = reader(files[0], self.chunks, **self._kwargs)
+        else:
+            self._ds = self._open_files(files)
+
+    def _get_schema(self):
+        """Make schema object, which embeds xarray object and some details"""
+        import xarray as xr
+        import msgpack
+        from .xarray_container import serialize_zarr_ds
+
+        self.urlpath, *_ = self._get_cache(self.urlpath)
+
+        if self._ds is None:
+            self._open_dataset()
+
+            ds2 = xr.Dataset({'raster': self._ds})
```

Comment:
`_ds` was an array but we want a data-set? I thought it should work anyway.

Reply:
This is copied directly from rasterio plugin. I think it needs to be a dataset to serialize to zarr?

Reply:
I am not sure, but if it works, that's fine. Does the word "raster" show up anywhere because of this?